datadog/dogstatsd/base.py

```diff
@@ -552,9 +552,11 @@ def _should_flush(self, length_to_be_added):
         return False
 
     def _flush_buffer(self):
-        self._send_to_server("\n".join(self.buffer))
-        self._current_buffer_total_size = 0
-        self.buffer = []
+        with self.lock:
```

Comment:
This lock is used for socket handling, it's not a great idea to re-use it for another purpose.

This also only solves a potential race around `_flush_buffer`, while the original bug report mentioned `open_buffer`  as it's the first one to be the issue.